tools/pubsys/src/aws/ami/mod.rs

```diff
@@ -0,0 +1,324 @@
+use snafu::{ensure, OptionExt, ResultExt};
+use std::collections::{HashMap, VecDeque};
+use std::path::PathBuf;
+use structopt::StructOpt;
+use wait::wait_for_ami;
+
+/// Builds Bottlerocket AMIs using latest build artifacts
+#[derive(Debug, StructOpt)]
+#[structopt(setting = clap::AppSettings::DeriveDisplayOrder)]
+pub(crate) struct AmiArgs {
+    /// Path to the image containing the root volume
+    #[structopt(short = "r", long, parse(from_os_str))]
+    root_image: PathBuf,
+
+    /// Path to the image containing the data volume
+    #[structopt(short = "d", long, parse(from_os_str))]
+    data_image: PathBuf,
+
+    /// Desired root volume size in gibibytes
+    #[structopt(long)]
+    root_volume_size: Option<i64>,
+
+    /// Desired data volume size in gibibytes
+    #[structopt(long)]
+    data_volume_size: i64,
+
+    /// The architecture of the machine image
+    #[structopt(short = "a", long)]
+    arch: String,
+
+    /// The desired AMI name
+    #[structopt(short = "n", long)]
+    name: String,
+
+    /// The desired AMI description
+    #[structopt(long)]
+    description: Option<String>,
+
+    /// Don't display progress bars
+    #[structopt(long)]
+    no_progress: bool,
+
+    /// Regions where you want the AMI, the first will be used as the base for copying
+    #[structopt(long, use_delimiter = true)]
+    regions: Vec<String>,
+}
+
+/// Common entrypoint from main()
+pub(crate) async fn run(args: &Args, ami_args: &AmiArgs) -> Result<()> {
+    info!(
+        "Using infra config from path: {}",
+        args.infra_config_path.display()
+    );
+    let infra_config = InfraConfig::from_path(&args.infra_config_path).context(error::Config)?;
+    trace!("Parsed infra config: {:?}", infra_config);
+
+    let aws = infra_config.aws.context(error::MissingConfig {
+        missing: "aws section",
+    })?;
+
+    // If the user gave an override list of regions, use that, otherwise use what's in the config.
+    let mut regions = if !ami_args.regions.is_empty() {
+        VecDeque::from(ami_args.regions.clone())
+    } else {
+        aws.regions.clone()
+    }
+    .into_iter()
+    .map(|name| region_from_string(&name, &aws))
+    .collect::<Result<VecDeque<Region>>>()?;
+
+    // We register in this base region first, then copy from there to any other regions.
+    let base_region = regions.pop_front().context(error::MissingConfig {
+        missing: "aws.regions",
+    })?;
+
+    // Build EBS client for snapshot management, and EC2 client for registration
+    let ebs_client = build_client::<EbsClient>(&base_region, &aws).context(error::Client {
+        client_type: "EBS",
+        region: base_region.name(),
+    })?;
+    let ec2_client = build_client::<Ec2Client>(&base_region, &aws).context(error::Client {
+        client_type: "EC2",
+        region: base_region.name(),
+    })?;
+
+    // Check if the AMI already exists, in which case we can use the existing ID, otherwise we
+    // register a new one.
```

Comment:
Is there a danger of accidentally copying an existing AMI when you meant to push? i.e. if you forgot to delete a previous bad push? (Maybe the `info!` below is enough...)

Reply:
It's only possible with `-dirty` AMIs, because the commit ID is included in the name, so if you add a commit (or even amend/rebase) you'll get new AMIs.  No one should be using `-dirty` AMIs for anything important...

Reply:
Makes sense, thanks!